# Notion d'axe

Pour une fonction, il est possible de préciser des axes pour préciser sur quels axes il faut effectuer l'opération.

![cube2.svg](cube2.svg)

In [11]:
import numpy as np
a = np.arange(5*4*7).reshape((5,4,7))
print(a)

[[[  0   1   2   3   4   5   6]
  [  7   8   9  10  11  12  13]
  [ 14  15  16  17  18  19  20]
  [ 21  22  23  24  25  26  27]]

 [[ 28  29  30  31  32  33  34]
  [ 35  36  37  38  39  40  41]
  [ 42  43  44  45  46  47  48]
  [ 49  50  51  52  53  54  55]]

 [[ 56  57  58  59  60  61  62]
  [ 63  64  65  66  67  68  69]
  [ 70  71  72  73  74  75  76]
  [ 77  78  79  80  81  82  83]]

 [[ 84  85  86  87  88  89  90]
  [ 91  92  93  94  95  96  97]
  [ 98  99 100 101 102 103 104]
  [105 106 107 108 109 110 111]]

 [[112 113 114 115 116 117 118]
  [119 120 121 122 123 124 125]
  [126 127 128 129 130 131 132]
  [133 134 135 136 137 138 139]]]


Pour la fonction somme, il est par exemple possible de l'appliquer à tout le tableau :

In [12]:
print(np.sum(a)) 

9730


On retrouve alors la formule de la somme des $n=139$ premiers entiers : 
    $$S = \dfrac{n(n+1)}{2} = 139\times30 = 9730$$

Mais on pourrait aussi vouloir faire une sommation uniquement selon le dernier axe (celui de shape 7), ou autrement dit selon une « ligne » du cube.

![lines2.svg](lines2.svg)

In [14]:
sum_last_axis = np.sum(a,axis = -1)
print(sum_last_axis.shape)
print(sum_last_axis)

(5, 4)
[[ 21  70 119 168]
 [217 266 315 364]
 [413 462 511 560]
 [609 658 707 756]
 [805 854 903 952]]


On se retrouve alors avec un nouveau tableau pour lequel on a sommé les éléments selon le dernier axe. Pour un cube, après avoir effectué cette sommation, on se retouve alors avec « seulement » une matrice et on a perdu une dimension.

Il est également possible de faire la sommation sur plusieurs axes (ici, selon un plan du cube) :

![planes2.svg](planes2.svg)

In [15]:
sum_last_axes = np.sum(a,axis = (1,2))
print(sum_last_axes.shape)
print(sum_last_axes)

(5,)
[ 378 1162 1946 2730 3514]


On se retouve alors avec seulement un vecteur qui correspond à la somme des éléments dans la ligne de `sum_last_axis`.